numba_cuda/numba/cuda/dispatcher.py

```diff
@@ -264,17 +265,24 @@ def maybe_link_nrt(self, link, tgt_ctx, asm):
         if not tgt_ctx.enable_nrt:
             return
 
-        all_nrt = "|".join(self.NRT_functions)
-        pattern = (
-            r"\.extern\s+\.func\s+(?:\s*\(.+\)\s*)?("
-            + all_nrt
-            + r")\s*\([^)]*\)\s*;"
-        )
-
-        nrt_in_asm = re.findall(pattern, asm)
-
-        basedir = os.path.dirname(os.path.abspath(__file__))
-        if nrt_in_asm:
+        nrt_in_asm = lambda asm: any(fn in asm for fn in self.NRT_functions)
+        link_nrt = nrt_in_asm(asm)
+        if not link_nrt:
+            for file in link:
+                if isinstance(file, LinkableCode):
+                    asm = file.data.decode("utf-8")
```

Comment:
`LinkableCode` can also be binary formats like archives, fatbins or LTOIR. Only text format is applicable here, which includes `CUSource` and `PTXSource`

```suggestion
                if isinstance(file,(CUSource, PTXSource)):
                    asm = file.data.decode("utf-8")
```

When it comes to other kind, we would have to silently not parse them I suppose? Disassembling them also works but would be too high a price to pay IMO.